streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -523,16 +528,21 @@ private boolean wrappedExceptionIsIn(final Throwable throwable, final Set<Class<
     }
 
     private void handleStreamsUncaughtException(final Throwable throwable,
-                                                final StreamsUncaughtExceptionHandler streamsUncaughtExceptionHandler) {
+                                                final StreamsUncaughtExceptionHandler streamsUncaughtExceptionHandler,
+                                                final boolean skipThreadReplacement) {
         final StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse action = getActionForThrowable(throwable, streamsUncaughtExceptionHandler);
         if (oldHandler) {
             log.warn("Stream's new uncaught exception handler is set as well as the deprecated old handler." +
                     "The old handler will be ignored as long as a new handler is set.");
         }
         switch (action) {
             case REPLACE_THREAD:
-                log.error("Replacing thread in the streams uncaught exception handler", throwable);
-                replaceStreamThread(throwable);
+                if (!skipThreadReplacement) {
+                    log.error("Replacing thread in the streams uncaught exception handler", throwable);
+                    replaceStreamThread(throwable);
+                } else {
+                    log.debug("Skipping thread replacement for recoverable error");
```

Comment:
I'd suggest logging it as a warn instead, also we can be more explicit in the message like "user decides to replace threads, which is not necessary; the library overrides that decision and skip.. please consider changing your exception handler class to not do so.."

Reply:
The thing is that the user _has_ to return something from the handler, there is no option to "do nothing" -- from the user perspective, the `REPLACE_THREAD` option essentially means "continue running the application", they don't care if we do so by literally replacing the thread or by just swallowing the exception. So imho it doesn't make sense to log as a warning because the user _should_ choose the `REPLACE_THREAD` option in this case.

I think the confusion is around the name "REPLACE_THREAD" when we start to consider optimizations where we don't need to actually replace the thread because we don't need to kill it in the first place, unfortunately we're stuck with the name unless we want to deprecate it and choose a better one (eg rename REPLACE_THREAD --> CONTINUE_RUNNING -- does that make sense?)

Reply:
I agree that `REPLACE_THREAD` leaks quite some internals. `CONTINUE` would be a better term. Actually, it doesn't matter what we do internally to ensure continuing the processing.

Reply:
+1 on `CONTINUE`, where the library could decide what to do actually, like whether we should replace the thread or not due to optimizations. I think it's not too late to replace it via a small KIP --- this does not need to block this PR though, just that the more I think of it, the more I feel we should make it more abstract away from implementation details.